### PR TITLE
Obj_Lightswitch (Sunswitch and Stonetower Temple flip switch)

### PR DESCRIPTION
**Pull request by isghj5**
Three non-equivelent functions:

The function that spawns burning flames for the fake light switches
The polyxlu draw function for the fake switches
The polyopa function for real switches

Both draw functions are nearly identical and have the same issue, but I couldn't figure out how to get them to match

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 1]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+                this->hitState = 0;
+            }
+            func_8095FFF8(this, func_8096012C, 1); // setup switch on transition
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/ObjLightswitch_Init.asm")
+// setup: trigger switch stage 1
+void func_8096012C(ObjLightswitch* this) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:244]: `ObjLightSwitch_SetupAsleep`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 2]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+    this->faceState = LIGHTSWITCH_FACE_ASLEEP;
+    this->colorShiftTimer = 0;
+    this->actionFunc = func_80960148;
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/ObjLightswitch_Destroy.asm")
+//action func: switch is triggered, color shift has begun
+void func_80960148(ObjLightswitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:251]: `ObjLightSwitch_Asleep`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 3]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+        func_80960224(this);
+    } else if (this->colorShiftTimer == 15) {
+        this->faceState = LIGHTSWITCH_FACE_WAKING;
+        Audio_PlayActorSound2(this, 0x2815); // sfx NA_SE_EV_FOOT_SWITCH
+    }
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_80960088.asm")
+// setup: switch is fully on
+void func_80960224(ObjLightswitch* this) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:271]: `ObjLightSwitch_SetupEnabled`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 4]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+    this->colorB = 0x3FC0;
+    this->colorAlpha = 0x3FC0;
+    this->edgeRotSpeed = -0xAA;
+    this->colorShiftTimer = 0;
+    this->actionFunc = func_80960260;
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_809600BC.asm")
+//action func: switch is fully on
+void func_80960260(ObjLightswitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:283]: `ObjLightSwitch_Enabled`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 5]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+        }
+    } else {
+        this->colorShiftTimer = 0;
+    }
+    this->edgeRot += this->edgeRotSpeed;
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_80960148.asm")
+// setup: disable switch transition
+void func_8096034C(ObjLightswitch* this) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:310]: `ObjLightSwitch_SetupDisabled`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 6]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+    this->colorShiftTimer = 20;
+    this->faceState = LIGHTSWITCH_FACE_WAKING;
+    this->actionFunc = func_80960370;
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_80960224.asm")
+// action func: disable switch transition
+void func_80960370(ObjLightswitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:317]: `ObjLightSwitch_Disabled`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 7]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+        func_80960088(this); // setup switch to off
+    } else if (this->colorShiftTimer == 15) {
+        this->faceState = LIGHTSWITCH_FACE_ASLEEP;
+        Audio_PlayActorSound2(&this->actor, 0x2815); // NA_SE_EV_FOOT_SWITCH
+    }
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_8096034C.asm")
+// setup: fake switch burn away
+void func_80960424(ObjLightswitch* this) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:333]: `ObjLightSwitch_SetupFade`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 8]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+    this->colorAlpha = 0x3FC0;
+    this->actionFunc = func_80960440;
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_80960370.asm")
+// action func: fake switch burning away
+void func_80960440(ObjLightswitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:339]: `ObjLightSwitch_Fade`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 9]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+    CLOSE_DISPS(globalCtx->state.gfxCtx);
+    
+}
+#else
 #pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Lightswitch_0x8095FBF0/func_809605F4.asm")
+#endif
+
+#if NON_EQUIVELENT
+// same issue as above
+void func_80960880(ObjLightswitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:454]: `ObjLightSwitch_DrawXlu`

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 10]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+        this->previousACFlags = this->collider.base.acFlags;
+        CollisionCheck_SetOC(globalCtx, &globalCtx->colCheckCtx, &this->collider);
+        CollisionCheck_SetAC(globalCtx, &globalCtx->colCheckCtx, &this->collider);
+    }
+}
+
+#if NON_EQUIVELENT
+// gDPSetEnvColor parameter load order is different than the order in which the value is built
+// also massive regalloc probably caused by the above
+void func_809605F4(ObjLightswitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:393]: `ObjLightSwitch_DrawOpa`

**Review by Kenix3 — CHANGES_REQUESTED** [review]
Names are suggested, but the naming convention and format are what is expected. Thank you for the submission!

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 11]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+};
+
+static ColliderJntSphInit sJntSphInit = {
+    { COLTYPE_NONE, AT_NONE, AC_ON | AC_TYPE_PLAYER | AC_TYPE_OTHER, OC1_ON | OC1_TYPE_ALL, OC2_TYPE_2, COLSHAPE_JNTSPH, },
+    1, sJntSphElementsInit,
+};
+
+// segmented addresses for poly opa and xlu funcs
+// different face addresses for sleep -> wake of light switch face
+void* D_80960BC4[] = { 0x06000C20, 0x06000420, 0x06001420,}; // D_80960BC4
```
[the remark this answers is not in the document]
Comment (isghj5) [reply src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:54]: These are going to be parts of the object right? gfx pointers?

Can I wait until we can decomp objects? There was a discussion about those being required at some point right?

**Comment by isghj5** [comment]
Hmm, I started implemented requested changes and now its giving a weird error on make (regular/clean and init)
```
cp: -r not specified; omitting directory 'baserom/assets'
make: *** [Makefile:200: build/baserom/assets] Error 1
```

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c [hunk 12]
```diff
@@ -21,46 +36,487 @@ const ActorInit Obj_Lightswitch_InitVars = {
...
+
+// colors used when spawning an effect
+Color_RGBA8 D_80960BD0 = { 0xFF, 0xFF, 0xA0, 0xA0, };
```
[the remark this answers is not in the document]
Comment (isghj5) [reply src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c:57]: I don't know how to rename these currently, as they are used in the non-matching polyopa/polyxlu functions the asm is looking for the old values. Do I have to re-enable asm data in code_script.txt?

File: src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (Kenix3) [reply src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c]: I would like to see symbols for all known addresses. I've been requiring all syms for assets, even in arrays.

Object decomp is when names for the display lists etc will be required. :) We are still waiting a bit for Tharo to be done with his semester at school to see if we can automate object decomp.

Should also add that the type for this looks to be `Gfx*`

**Review by Kenix3 — CHANGES_REQUESTED** [review]
Sorry again, but I noticed a few more things.

**Review by Kenix3 — APPROVED** [review]
Looks good to me, this is ready to merge.

**Pull request merged**